### ISSUE TITLE
TLS Client Authentication only on CERTIFICATE_REQUEST?

### ISSUE THREAD

**Issue by reneme** [randombit/botan#1925]
After readying our [TLS::Stream](https://github.com/randombit/botan/pull/1839) PR, we are now busy integrating it into our application.
One use-case is to do a login via a TLS client certificate. On our production server we receive the callback `find_cert_chain()` just fine and the authentication succeeds. On the development installation it doesn't work though, we don't even receive the mentioned callback.

Now, looking at [this code](https://github.com/randombit/botan/blob/master/src/lib/tls/tls_client.cpp#L523) in tls_client.cpp it seems as if a `CERTIFICATE_REQUEST` state is hard-required for Botan to even consider to send a client certificate. Both OpenSSL and Microsoft's SChannel seem to not care and use a provided client certificate in any case.

A quick glance with `openssl s_client` seems to verify that our dev installation does not send such a `CERTIFICATE_REQUEST`. Is that something that needs fixing in Botan's TLS implementation?

## On Production

```
> openssl s_client -connect prod-host.com:443 -state
CONNECTED(00000005)
SSL_connect:before/connect initialization
SSL_connect:SSLv3 write client hello A
SSL_connect:SSLv3 read server hello A
depth=2 C = DE, O = D-Trust GmbH, CN = D-TRUST Root Class 3 CA 2 EV 2009
verify error:num=19:self signed certificate in certificate chain
verify return:0
SSL_connect:SSLv3 read server certificate A
SSL_connect:SSLv3 read server key exchange A
SSL_connect:SSLv3 read server certificate request A   <<== this is missing in our dev installation
SSL_connect:SSLv3 read server done A
SSL_connect:SSLv3 write client certificate A
SSL_connect:SSLv3 write client key exchange A
SSL_connect:SSLv3 write change cipher spec A
SSL_connect:SSLv3 write finished A
SSL_connect:SSLv3 flush data
SSL_connect:SSLv3 read finished A
[... snip ...]
```

## On Development

```
> openssl s_client -connect dev-host.com:443 -state
CONNECTED(00000005)
SSL_connect:before/connect initialization
SSL_connect:SSLv3 write client hello A
SSL_connect:SSLv3 read server hello A
depth=2 O = Digital Signature Trust Co., CN = DST Root CA X3
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
verify return:1
depth=0 CN = nexenio.space
verify return:1
SSL_connect:SSLv3 read server certificate A
SSL_connect:SSLv3 read server key exchange A
SSL_connect:SSLv3 read server done A
SSL_connect:SSLv3 write client key exchange A
SSL_connect:SSLv3 write change cipher spec A
SSL_connect:SSLv3 write finished A
SSL_connect:SSLv3 flush data
SSL_connect:SSLv3 read server session ticket A
SSL_connect:SSLv3 read finished A
[... snip ...]
```

### Potentially Related Issues

* [How to write TLS code with client certificate? #282](https://github.com/randombit/botan/issues/282)

**Comment by randombit** [comment]
Wait, so OpenSSL if configured with a client cert will send it *even if the server does not ask for it*? That seems really strange to me. And from a quick look contradicts https://tools.ietf.org/html/rfc5246#section-7.4.6

"This [Client Certificate] message is **only sent if the server requests a certificate**.  If no suitable certificate is available, the client MUST send a certificate message containing no certificates."

I would say if possible your dev env should request a cert if it wants one. It's certainly possible something needs to change in how we handle this but the above text seems pretty clear.

**Comment by seppy87** [comment]
Is Botan built on OpenSSL or an independent SSL Library?

**Comment by reneme** [comment]
> Is Botan built on OpenSSL or an independent SSL Library?

To answer your question: Botan implements TLS independently from OpenSSL, thats what we are trying to use.

### TL;DR: I was wrong, we had a bug in our application and a funny configuration that sidetracked us.

In detail: Our dev installation has a load balancer that manages all endpoints of our backend infrastructure. One of those endpoints requires a client authentication via TLS. So the load balancer relies on the "Server Name Information" to select the TLS configuration to use.

While prototypically integrating the new [TLS Stream](https://github.com/randombit/botan/pull/1839) into [cpprestsdk](https://github.com/Microsoft/cpprestsdk) (our HTTP library right now), we screwed up the configuration of the SNI. Hence, our load balancer (probably) defaulted to *not* requiring a  client certificate. The production setup is different in that respect and worked regardlessly.

Now, when trying to reproduce the issue with `openssl` on the macOS command line, it produced the output I initially posted: i.e. no `CERTIFICATE_REQUEST` was received from the server. When running our application based on OpenSSL's TLS, it worked flawlessly on the dev installation. This provoked me to open this issue.

However, TIL: macOS's stock `openssl` CLI tool seems to be "LibreSSL" actually. 🤦‍♂
```
$ openssl version
LibreSSL 2.6.5
```

When trying the `s_client` call with (the actual) OpenSSL 1.1 CLI, everything works fine and the list of acceptable CAs is received from the server as expected. That happens, because OpenSSL 1.1 automagically adds the SNI to the ClientHello, while LibreSSL requires an extra CLI flag for that.

Bottom line: We fixed our cpprestsdk integration code and it works fine now. Sorry for the noise.
